keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,630 @@
+
+```golang
+type ServiceReference struct {
+    ...
+
+     // If specified, the port on the service that hosting webhook.
+     // Default to 443 for backward compatibility.
+     // +optional
+     Port *int32 `json:"port,omitempty" protobuf:"varint,4,opt,name=port"`
+}
+```
+
+### Plumbing existing objects to delete admission
+
+Current admission webhooks can hook on `DELETE` and `DELETE_COLLECTION` events but they won't get any object in `oldObject` or `Object` field of `AdmissionRequest`. The proposal is to send them existing object(s) as the `oldObject`. This is also helpful for applying `ObjectSelector` to these webhooks. Note that `oldObject` for `DELETE_COLLECTION` will be a `v1.list` of existing objects but the list may not have all items. API server may call the webhook multiple times with sub-lists. There is no API change for this feature, only documentation:
+
+```golang
+type AdmissionRequest struct {
+  ...
+  // OldObject is the existing object. Only populated for UPDATE, DELETE, and DELETE_COLLECTION requests.
+  // +optional
+  OldObject runtime.RawExtension `json:"oldObject,omitempty" protobuf:"bytes,10,opt,name=oldObject"`
+  ..
+}
+```
+
+### Mutating Plugin ordering
+
+Current ordering of the plugins (including webhooks) is fixed but may not work for all cases (e.g. [this issue](https://github.com/kubernetes/kubernetes/issues/64333)). A mutating webhook can add a completely new structure to the object (e.g. a `Container`) and other mutating plugins may have opinion on those new structures (e.g. setting the image pull policy on all containers). Although this problem can go deeper than two level (if the second mutating webhook added a new structure based on the new `Container`), with current set of standard mutating plugins and use-cases, it look like a rerun of mutating plugins is a sufficient fix. The proposal is to rerun all mutating plugins (including webhooks) if there is any change by mutating webhooks. The assumption for this process is all mutating plugins are idempotent and this must be clearly documented to users.
+Mutations from in-tree plugins will not trigger this process as they are well-ordered but if there is any mutation by webhooks, all of the plugins including in-tree ones will be run again.
+
+### Passing {Operation}Option to Webhook
+
+Each of the operations webhook can have an `Option` structure (e.g. `DeleteOption` or `CreateOption`) passed by the user. It is useful for some webhooks to know what are those options user requested to better modify or validate object. The proposal is to add those Options as an `OperationOption` field to the `AdmissionRequest` API object.
```

Comment:
Describe what will be done with ConnectOptions for CONNECT operations which are currently passed as the main object (and are possibly able to be mutated?). Will those be moved into the options field or left where they are in v1beta1? in v1?

Reply:
for compatibility, we should not change the behavior in v1beta1 AdmissionReviews. An open question for v1 AdmissionReviews sgtm